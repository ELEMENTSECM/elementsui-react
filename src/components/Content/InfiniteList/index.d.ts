import * as React from 'react';

export interface InfiniteListProps {
	/**
	 * DOM element's id attribute
	 */
	id?: string;
	/**
	 * Locale (en, nb, etc.)
	 */
	locale?: string;
	/**
	 * A function which must be called after reaching the bottom. It must trigger some sort of action which fetches the next data. The data is passed as children to the InfiniteList component and the data should contain previous items too. e.g. Initial data = [1, 2, 3] and then next load of data should be [1, 2, 3, 4, 5, 6].
	 */
	next?: (...args: any[]) => any;
	/**
	 * It tells the InfiniteList component on whether to call next function on reaching the bottom and shows an endMessage to the user
	 */
	hasMore?: boolean;
	/**
	 * The data items which you need to scroll.
	 */
	children?: React.ReactNode;
	/**
	 * This message is shown to the user when he has seen all the records which means he's at the bottom and hasMore is false
	 */
	endMessage?: React.ReactNode;
	/**
	 * Optional, give only if you want to have a fixed height scrolling content
	 */
	height?: number;
	/**
	 * Optional, reference to a (parent) DOM element that is already providing overflow scrollbars to the InfiniteList component. You should provide the id of the DOM node preferably.
	 */
	scrollableTarget?: React.ReactNode;
	/**
	 * Children is by default assumed to be of type array and it's length is used to determine if loader needs to be shown or not, if your children is not an array, specify this prop to tell if your items are 0 or more.
	 */
	hasChildren?: boolean;
	/**
	 * To enable Pull Down to Refresh feature
	 */
	pullDownToRefresh?: boolean;
	/**
	 * Any JSX that you want to show the user
	 */
	pullDownToRefreshContent?: React.ReactNode;
	/**
	 * Any JSX that you want to show the user
	 */
	releaseToRefreshContent?: React.ReactNode;
	/**
	 * Minimum distance the user needs to pull down to trigger the refresh
	 */
	pullDownToRefreshThreshold?: number;
	/**
	 * This function will be called, it should return the fresh data that you want to show the user
	 */
	refreshFunction?: (...args: any[]) => any;
	/**
	 * A function that will listen to the scroll event on the scrolling container. Note that the scroll event is throttled, so you may not receive as many events as you would expect.
	 */
	onScroll?: (...args: any[]) => any;
	/**
	 * Set the length of the data.This will unlock the subsequent calls to next.
	 */
	dataLength: number;
	/**
	 * User-defined styling
	 */
	styles?: (...args: any[]) => any;
}

declare const InfiniteList: React.SFC<InfiniteListProps>;

export default InfiniteList;
