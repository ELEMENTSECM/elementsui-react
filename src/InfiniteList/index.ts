export { default, InfiniteListProps } from "./InfiniteList";
export { default as ListItem, InfiniteListItemProps } from "./InfiniteListItem";
