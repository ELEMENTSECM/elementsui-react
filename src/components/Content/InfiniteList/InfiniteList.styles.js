export const styles = props => {
	const { className } = props;

	return {
		root: [
			'InfiniteList',
			{
				overflow: 'auto',
				WebkitOverflowScrolling: 'touch'
			},
			className
		],
		pulldown: [{ position: 'relative' }],
		pulldownHandle: [
			{
				position: 'absolute',
				left: 0,
				right: 0
			}
		],
		list: []
	};
};
