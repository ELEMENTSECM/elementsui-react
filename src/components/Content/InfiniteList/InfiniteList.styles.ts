import { css } from "react-emotion";

export const styles = {
	InfiniteList: css`
		overflow: auto;
		-webkit-overflow-scrolling: touch;
	`,
	pulldown: css`position: relative;`,
	pulldownHandle: css`
		position: absolute;
		,: 0;
		right: 0;
	`
};
