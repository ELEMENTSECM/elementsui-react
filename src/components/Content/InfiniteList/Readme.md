InfiniteList

```js
<InfiniteList locale="en" dataLength={20} hasMore={false}>
	{[ ...Array(20) ].map((_, i) => <div key={i}>Item {i}</div>)}
</InfiniteList>
```