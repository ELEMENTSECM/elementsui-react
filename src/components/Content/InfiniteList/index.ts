export { default, IInfiniteListProps } from "./InfiniteList";
