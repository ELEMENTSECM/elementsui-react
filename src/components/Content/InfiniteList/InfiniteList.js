import * as React from 'react';
import PropTypes from 'prop-types';
import { styles } from './InfiniteList.styles';
import { classNamesFunction, customizable, styled } from 'office-ui-fabric-react/lib/Utilities';
import Spinner from '../../Indicators/Spinner';
import { IntlProvider, FormattedMessage } from 'react-intl';
import nls from './InfiniteList.nls.json';

/** InfiniteList example */
class InfiniteListComponent extends React.Component {
	constructor(props) {
		super();
		this.state = {
			showLoader: false,
			lastScrollTop: 0,
			actionTriggered: false,
			pullToRefreshThresholdBreached: false
		};
		// variables to keep track of pull down behaviour
		this.startY = 0;
		this.currentY = 0;
		this.dragging = false;
		// will be populated in componentDidMount
		// based on the height of the pull down element
		this.maxPullDownDistance = 0;

		this.onScrollListener = this.onScrollListener.bind(this);
		this.throttledOnScrollListener = this.throttle(this.onScrollListener, 150).bind(this);
		this.onStart = this.onStart.bind(this);
		this.onMove = this.onMove.bind(this);
		this.onEnd = this.onEnd.bind(this);
		this.getScrollableTarget = this.getScrollableTarget.bind(this);
		this.classNames = classNamesFunction()(styles, props);
	}

	componentDidMount() {
		this._scrollableNode = this.getScrollableTarget();
		this.el = this.props.height ? this._infScroll : this._scrollableNode || window;
		this.el.addEventListener('scroll', this.throttledOnScrollListener);

		if (
			typeof this.props.initialScrollY === 'number' &&
			this.el.scrollHeight > this.props.initialScrollY
		) {
			this.el.scrollTo(0, this.props.initialScrollY);
		}

		if (this.props.pullDownToRefresh) {
			this.el.addEventListener('touchstart', this.onStart);
			this.el.addEventListener('touchmove', this.onMove);
			this.el.addEventListener('touchend', this.onEnd);

			this.el.addEventListener('mousedown', this.onStart);
			this.el.addEventListener('mousemove', this.onMove);
			this.el.addEventListener('mouseup', this.onEnd);

			// get BCR of pullDown element to position it above
			this.maxPullDownDistance = this._pullDown.firstChild.getBoundingClientRect().height;
			this.forceUpdate();

			if (typeof this.props.refreshFunction !== 'function') {
				throw new Error(
					`Mandatory prop "refreshFunction" missing.
			  Pull Down To Refresh functionality will not work
			  as expected. Check README.md for usage'`
				);
			}
		}
	}

	componentWillUnmount() {
		this.el.removeEventListener('scroll', this.throttledOnScrollListener);

		if (this.props.pullDownToRefresh) {
			this.el.removeEventListener('touchstart', this.onStart);
			this.el.removeEventListener('touchmove', this.onMove);
			this.el.removeEventListener('touchend', this.onEnd);

			this.el.removeEventListener('mousedown', this.onStart);
			this.el.removeEventListener('mousemove', this.onMove);
			this.el.removeEventListener('mouseup', this.onEnd);
		}
	}

	componentWillReceiveProps(props) {
		// do nothing when dataLength is unchanged
		if (this.props.dataLength === props.dataLength) return;

		// update state when new data was sent in
		this.setState({
			showLoader: false,
			actionTriggered: false,
			pullToRefreshThresholdBreached: false
		});
	}

	getScrollableTarget() {
		if (this.props.scrollableTarget instanceof HTMLElement) return this.props.scrollableTarget;
		if (typeof this.props.scrollableTarget === 'string') {
			return document.getElementById(this.props.scrollableTarget);
		}
		if (this.props.scrollableTarget === null) {
			console.warn(`You are trying to pass scrollableTarget but it is null. This might
			happen because the element may not have been added to DOM yet.
			See https://github.com/ankeetmaini/react-infinite-scroll-component/issues/59 for more info.
		  `);
		}
		return null;
	}

	onStart(evt) {
		if (this.state.lastScrollTop) return;

		this.dragging = true;
		this.startY = evt.pageY || evt.touches[0].pageY;
		this.currentY = this.startY;

		this._infScroll.style.willChange = 'transform';
		this._infScroll.style.transition = `transform 0.2s cubic-bezier(0,0,0.31,1)`;
	}

	onMove(evt) {
		if (!this.dragging) return;
		this.currentY = evt.pageY || evt.touches[0].pageY;

		// user is scrolling down to up
		if (this.currentY < this.startY) return;

		const threshold = this.props.pullDownToRefreshThreshold || 100;
		if (this.currentY - this.startY >= threshold) {
			this.setState({
				pullToRefreshThresholdBreached: true
			});
		}

		// so you can drag upto 1.5 times of the maxPullDownDistance
		if (this.currentY - this.startY > this.maxPullDownDistance * 1.5) return;

		this._infScroll.style.overflow = 'visible';
		this._infScroll.style.transform = `translate3d(0px, ${this.currentY - this.startY}px, 0px)`;
	}

	onEnd(evt) {
		this.startY = 0;
		this.currentY = 0;

		this.dragging = false;

		if (this.state.pullToRefreshThresholdBreached) {
			this.props.refreshFunction && this.props.refreshFunction();
		}

		requestAnimationFrame(() => {
			this._infScroll.style.overflow = 'auto';
			this._infScroll.style.transform = 'none';
			this._infScroll.style.willChange = 'none';
		});
	}

	isElementAtBottom(target, scrollThreshold = 0.8) {
		const clientHeight =
			target === document.body || target === document.documentElement
				? window.screen.availHeight
				: target.clientHeight;

		return target.scrollTop + clientHeight >= scrollThreshold * target.scrollHeight;
	}

	onScrollListener(event) {
		if (typeof this.props.onScroll === 'function') {
			// Execute this callback in next tick so that it does not affect the
			// functionality of the library.
			setTimeout(() => this.props.onScroll(event), 0);
		}

		let target =
			this.props.height || this._scrollableNode
				? event.target
				: document.documentElement.scrollTop
					? document.documentElement
					: document.body;

		// return immediately if the action has already been triggered,
		// prevents multiple triggers.
		if (this.state.actionTriggered) return;

		let atBottom = this.isElementAtBottom(target, this.props.scrollThreshold);

		// call the `next` function in the props to trigger the next data fetch
		if (atBottom && this.props.hasMore) {
			this.setState({ actionTriggered: true, showLoader: true });
			this.props.next();
		}
		this.setState({ lastScrollTop: target.scrollTop });
	}

	throttle(fn, threshhold, scope) {
		threshhold || (threshhold = 250);
		var last, deferTimer;
		return function() {
			var context = scope || this;

			var now = +new Date(),
				args = arguments;
			if (last && now < last + threshhold) {
				clearTimeout(deferTimer);
				deferTimer = setTimeout(function() {
					last = now;
					fn.apply(context, args);
				}, threshhold);
			} else {
				last = now;
				fn.apply(context, args);
			}
		};
	}

	render() {
		const style = {
			height: this.props.height || 'auto'
		};
		const hasChildren = this.props.hasChildren || !!(this.props.children && this.props.children.length);

		const outerDivStyle = this.props.pullDownToRefresh && this.props.height ? { overflow: 'auto' } : {};
		const { root, pulldown, list, pulldownHandle } = this.classNames;
		return (
			<IntlProvider locale={this.props.locale} messages={nls[this.props.locale]}>
				<div style={outerDivStyle} className={root}>
					<div ref={infScroll => (this._infScroll = infScroll)} style={style}>
						{this.props.pullDownToRefresh && (
							<div className={pulldown} ref={pullDown => (this._pullDown = pullDown)}>
								<div
									className={pulldownHandle}
									style={{ top: -1 * this.maxPullDownDistance }}>
									{!this.state.pullToRefreshThresholdBreached && (
										<h3>
											<FormattedMessage id="pullDownToRefresh" />
										</h3>
									)}
									{this.state.pullToRefreshThresholdBreached && (
										<h3>
											<FormattedMessage id="releaseToRefresh" />
										</h3>
									)}
								</div>
							</div>
						)}
						<ul className={list}>{this.props.children}</ul>
						{!this.state.showLoader && !hasChildren && this.props.hasMore && <Spinner />}
						{this.state.showLoader && this.props.hasMore && <Spinner />}
						{!this.props.hasMore && this.props.endMessage}
					</div>
				</div>
			</IntlProvider>
		);
	}
}

export const InfiniteList = InfiniteListComponent;

InfiniteList.propTypes = {
	/** Locale (en, nb, etc.) */
	locale: PropTypes.string,
	/** A function which must be called after reaching the bottom. It must trigger some sort of action which fetches the next data. The data is passed as children to the InfiniteList component and the data should contain previous items too. e.g. Initial data = [1, 2, 3] and then next load of data should be [1, 2, 3, 4, 5, 6]. */
	next: PropTypes.func,
	/** It tells the InfiniteList component on whether to call next function on reaching the bottom and shows an endMessage to the user */
	hasMore: PropTypes.bool,
	/** The data items which you need to scroll. */
	children: PropTypes.node,
	/** A threshold value after that the InfiniteList will call next. By default it's 0.8. It means the next will be called when the user comes below 80% of the total height. */
	scrollThreshold: PropTypes.number,
	/** This message is shown to the user when he has seen all the records which means he's at the bottom and hasMore is false */
	endMessage: PropTypes.node,
	/** Optional, give only if you want to have a fixed height scrolling content */
	height: PropTypes.number,
	/** Optional, reference to a (parent) DOM element that is already providing overflow scrollbars to the InfiniteList component. You should provide the id of the DOM node preferably. */
	scrollableTarget: PropTypes.node,
	/** Children is by default assumed to be of type array and it's length is used to determine if loader needs to be shown or not, if your children is not an array, specify this prop to tell if your items are 0 or more. */
	hasChildren: PropTypes.bool,
	/** To enable Pull Down to Refresh feature */
	pullDownToRefresh: PropTypes.bool,
	/** Any JSX that you want to show the user */
	pullDownToRefreshContent: PropTypes.node,
	/** Any JSX that you want to show the user */
	releaseToRefreshContent: PropTypes.node,
	/** Minimum distance the user needs to pull down to trigger the refresh */
	pullDownToRefreshThreshold: PropTypes.number,
	/** This function will be called, it should return the fresh data that you want to show the user */
	refreshFunction: PropTypes.func,
	/** A function that will listen to the scroll event on the scrolling container. Note that the scroll event is throttled, so you may not receive as many events as you would expect. */
	onScroll: PropTypes.func,
	/** Set the length of the data.This will unlock the subsequent calls to next. */
	dataLength: PropTypes.number.isRequired,
	/** User-defined styling */
	styles: PropTypes.func
};

InfiniteListComponent.propTypes = {
	/** Locale (en, nb, etc.) */
	locale: PropTypes.string.isRequired,
	/** A function which must be called after reaching the bottom. It must trigger some sort of action which fetches the next data. The data is passed as children to the InfiniteList component and the data should contain previous items too. e.g. Initial data = [1, 2, 3] and then next load of data should be [1, 2, 3, 4, 5, 6]. */
	next: PropTypes.func,
	/** It tells the InfiniteList component on whether to call next function on reaching the bottom and shows an endMessage to the user */
	hasMore: PropTypes.bool,
	/** The data items which you need to scroll. */
	children: PropTypes.node,
	/** A threshold value after that the InfiniteList will call next. By default it's 0.8. It means the next will be called when the user comes below 80% of the total height. */
	scrollThreshold: PropTypes.number,
	/** This message is shown to the user when he has seen all the records which means he's at the bottom and hasMore is false */
	endMessage: PropTypes.node,
	/** Optional, give only if you want to have a fixed height scrolling content */
	height: PropTypes.number,
	/** Optional, reference to a (parent) DOM element that is already providing overflow scrollbars to the InfiniteList component. You should provide the id of the DOM node preferably. */
	scrollableTarget: PropTypes.node,
	/** Children is by default assumed to be of type array and it's length is used to determine if loader needs to be shown or not, if your children is not an array, specify this prop to tell if your items are 0 or more. */
	hasChildren: PropTypes.bool,
	/** To enable Pull Down to Refresh feature */
	pullDownToRefresh: PropTypes.bool,
	/** Any JSX that you want to show the user */
	pullDownToRefreshContent: PropTypes.node,
	/** Any JSX that you want to show the user */
	releaseToRefreshContent: PropTypes.node,
	/** Minimum distance the user needs to pull down to trigger the refresh */
	pullDownToRefreshThreshold: PropTypes.number,
	/** This function will be called, it should return the fresh data that you want to show the user */
	refreshFunction: PropTypes.func,
	/** A function that will listen to the scroll event on the scrolling container. Note that the scroll event is throttled, so you may not receive as many events as you would expect. */
	onScroll: PropTypes.func,
	/** Set the length of the data.This will unlock the subsequent calls to next. */
	dataLength: PropTypes.number.isRequired,
	/** User-defined styling */
	styles: PropTypes.func
};

export default styled(customizable('InfiniteList', ['theme'])(InfiniteList), styles);
