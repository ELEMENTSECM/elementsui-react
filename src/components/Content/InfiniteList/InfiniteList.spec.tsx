import * as React from "react";
import { shallow } from "enzyme";
import InfiniteList from "./index";

describe("InfiniteList", () => {
	test("displays correctly", () => {
		const wrapper = shallow(
			<InfiniteList locale="en" dataLength={20} hasMore={false}>
				{[ ...Array(20) ].map((_, i) => {
					return <div key={i}>Item {i}</div>;
				})}
			</InfiniteList>
		);
		expect(wrapper).toMatchSnapshot();
	});
});
