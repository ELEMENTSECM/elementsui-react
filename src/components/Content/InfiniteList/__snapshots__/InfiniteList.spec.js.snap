// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InfiniteList displays correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <StyledCustomizedInfiniteList
    dataLength={20}
    hasMore={false}
    locale="en"
  >
    <div>
      Item 
      0
    </div>
    <div>
      Item 
      1
    </div>
    <div>
      Item 
      2
    </div>
    <div>
      Item 
      3
    </div>
    <div>
      Item 
      4
    </div>
    <div>
      Item 
      5
    </div>
    <div>
      Item 
      6
    </div>
    <div>
      Item 
      7
    </div>
    <div>
      Item 
      8
    </div>
    <div>
      Item 
      9
    </div>
    <div>
      Item 
      10
    </div>
    <div>
      Item 
      11
    </div>
    <div>
      Item 
      12
    </div>
    <div>
      Item 
      13
    </div>
    <div>
      Item 
      14
    </div>
    <div>
      Item 
      15
    </div>
    <div>
      Item 
      16
    </div>
    <div>
      Item 
      17
    </div>
    <div>
      Item 
      18
    </div>
    <div>
      Item 
      19
    </div>
  </StyledCustomizedInfiniteList>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <div>
          Item 
          0
        </div>,
        <div>
          Item 
          1
        </div>,
        <div>
          Item 
          2
        </div>,
        <div>
          Item 
          3
        </div>,
        <div>
          Item 
          4
        </div>,
        <div>
          Item 
          5
        </div>,
        <div>
          Item 
          6
        </div>,
        <div>
          Item 
          7
        </div>,
        <div>
          Item 
          8
        </div>,
        <div>
          Item 
          9
        </div>,
        <div>
          Item 
          10
        </div>,
        <div>
          Item 
          11
        </div>,
        <div>
          Item 
          12
        </div>,
        <div>
          Item 
          13
        </div>,
        <div>
          Item 
          14
        </div>,
        <div>
          Item 
          15
        </div>,
        <div>
          Item 
          16
        </div>,
        <div>
          Item 
          17
        </div>,
        <div>
          Item 
          18
        </div>,
        <div>
          Item 
          19
        </div>,
      ],
      "dataLength": 20,
      "hasMore": false,
      "locale": "en",
      "styles": [Function],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "0",
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Item ",
            0,
          ],
        },
        "ref": null,
        "rendered": Array [
          "Item ",
          0,
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": "1",
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Item ",
            1,
          ],
        },
        "ref": null,
        "rendered": Array [
          "Item ",
          1,
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": "2",
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Item ",
            2,
          ],
        },
        "ref": null,
        "rendered": Array [
          "Item ",
          2,
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": "3",
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Item ",
            3,
          ],
        },
        "ref": null,
        "rendered": Array [
          "Item ",
          3,
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": "4",
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Item ",
            4,
          ],
        },
        "ref": null,
        "rendered": Array [
          "Item ",
          4,
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": "5",
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Item ",
            5,
          ],
        },
        "ref": null,
        "rendered": Array [
          "Item ",
          5,
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": "6",
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Item ",
            6,
          ],
        },
        "ref": null,
        "rendered": Array [
          "Item ",
          6,
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": "7",
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Item ",
            7,
          ],
        },
        "ref": null,
        "rendered": Array [
          "Item ",
          7,
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": "8",
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Item ",
            8,
          ],
        },
        "ref": null,
        "rendered": Array [
          "Item ",
          8,
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": "9",
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Item ",
            9,
          ],
        },
        "ref": null,
        "rendered": Array [
          "Item ",
          9,
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": "10",
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Item ",
            10,
          ],
        },
        "ref": null,
        "rendered": Array [
          "Item ",
          10,
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": "11",
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Item ",
            11,
          ],
        },
        "ref": null,
        "rendered": Array [
          "Item ",
          11,
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": "12",
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Item ",
            12,
          ],
        },
        "ref": null,
        "rendered": Array [
          "Item ",
          12,
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": "13",
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Item ",
            13,
          ],
        },
        "ref": null,
        "rendered": Array [
          "Item ",
          13,
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": "14",
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Item ",
            14,
          ],
        },
        "ref": null,
        "rendered": Array [
          "Item ",
          14,
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": "15",
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Item ",
            15,
          ],
        },
        "ref": null,
        "rendered": Array [
          "Item ",
          15,
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": "16",
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Item ",
            16,
          ],
        },
        "ref": null,
        "rendered": Array [
          "Item ",
          16,
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": "17",
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Item ",
            17,
          ],
        },
        "ref": null,
        "rendered": Array [
          "Item ",
          17,
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": "18",
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Item ",
            18,
          ],
        },
        "ref": null,
        "rendered": Array [
          "Item ",
          18,
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": "19",
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Item ",
            19,
          ],
        },
        "ref": null,
        "rendered": Array [
          "Item ",
          19,
        ],
        "type": "div",
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <div>
            Item 
            0
          </div>,
          <div>
            Item 
            1
          </div>,
          <div>
            Item 
            2
          </div>,
          <div>
            Item 
            3
          </div>,
          <div>
            Item 
            4
          </div>,
          <div>
            Item 
            5
          </div>,
          <div>
            Item 
            6
          </div>,
          <div>
            Item 
            7
          </div>,
          <div>
            Item 
            8
          </div>,
          <div>
            Item 
            9
          </div>,
          <div>
            Item 
            10
          </div>,
          <div>
            Item 
            11
          </div>,
          <div>
            Item 
            12
          </div>,
          <div>
            Item 
            13
          </div>,
          <div>
            Item 
            14
          </div>,
          <div>
            Item 
            15
          </div>,
          <div>
            Item 
            16
          </div>,
          <div>
            Item 
            17
          </div>,
          <div>
            Item 
            18
          </div>,
          <div>
            Item 
            19
          </div>,
        ],
        "dataLength": 20,
        "hasMore": false,
        "locale": "en",
        "styles": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "0",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Item ",
              0,
            ],
          },
          "ref": null,
          "rendered": Array [
            "Item ",
            0,
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": "1",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Item ",
              1,
            ],
          },
          "ref": null,
          "rendered": Array [
            "Item ",
            1,
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": "2",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Item ",
              2,
            ],
          },
          "ref": null,
          "rendered": Array [
            "Item ",
            2,
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": "3",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Item ",
              3,
            ],
          },
          "ref": null,
          "rendered": Array [
            "Item ",
            3,
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": "4",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Item ",
              4,
            ],
          },
          "ref": null,
          "rendered": Array [
            "Item ",
            4,
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": "5",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Item ",
              5,
            ],
          },
          "ref": null,
          "rendered": Array [
            "Item ",
            5,
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": "6",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Item ",
              6,
            ],
          },
          "ref": null,
          "rendered": Array [
            "Item ",
            6,
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": "7",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Item ",
              7,
            ],
          },
          "ref": null,
          "rendered": Array [
            "Item ",
            7,
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": "8",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Item ",
              8,
            ],
          },
          "ref": null,
          "rendered": Array [
            "Item ",
            8,
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": "9",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Item ",
              9,
            ],
          },
          "ref": null,
          "rendered": Array [
            "Item ",
            9,
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": "10",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Item ",
              10,
            ],
          },
          "ref": null,
          "rendered": Array [
            "Item ",
            10,
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": "11",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Item ",
              11,
            ],
          },
          "ref": null,
          "rendered": Array [
            "Item ",
            11,
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": "12",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Item ",
              12,
            ],
          },
          "ref": null,
          "rendered": Array [
            "Item ",
            12,
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": "13",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Item ",
              13,
            ],
          },
          "ref": null,
          "rendered": Array [
            "Item ",
            13,
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": "14",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Item ",
              14,
            ],
          },
          "ref": null,
          "rendered": Array [
            "Item ",
            14,
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": "15",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Item ",
              15,
            ],
          },
          "ref": null,
          "rendered": Array [
            "Item ",
            15,
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": "16",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Item ",
              16,
            ],
          },
          "ref": null,
          "rendered": Array [
            "Item ",
            16,
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": "17",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Item ",
              17,
            ],
          },
          "ref": null,
          "rendered": Array [
            "Item ",
            17,
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": "18",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Item ",
              18,
            ],
          },
          "ref": null,
          "rendered": Array [
            "Item ",
            18,
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": "19",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Item ",
              19,
            ],
          },
          "ref": null,
          "rendered": Array [
            "Item ",
            19,
          ],
          "type": "div",
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
